pkg/router/controller/host_admitter.go

```diff
@@ -0,0 +1,256 @@
+// admit routes and relay them to the next plugin in the chain.
+// Recorder is an interface for indicating why a route was rejected.
+func NewHostAdmitter(plugin router.Plugin, fn RouteAdmissionFunc, restrict bool, recorder RejectionRecorder) *HostAdmitter {
+	return &HostAdmitter{
+		plugin:   plugin,
+		admitter: fn,
+		recorder: recorder,
+
+		restrictOwnership: restrict,
+		claimedHosts:      RouteMap{},
+		claimedWildcards:  RouteMap{},
+		blockedWildcards:  RouteMap{},
+	}
+}
+
+// HandleNode processes watch events on the Node resource.
+func (p *HostAdmitter) HandleNode(eventType watch.EventType, node *kapi.Node) error {
+	return p.plugin.HandleNode(eventType, node)
+}
+
+// HandleEndpoints processes watch events on the Endpoints resource.
+func (p *HostAdmitter) HandleEndpoints(eventType watch.EventType, endpoints *kapi.Endpoints) error {
+	return p.plugin.HandleEndpoints(eventType, endpoints)
+}
+
+// HandleRoute processes watch events on the Route resource.
+func (p *HostAdmitter) HandleRoute(eventType watch.EventType, route *routeapi.Route) error {
+	err, allow := p.admitter(route)
+	if !allow {
+		msg := "blocked by admitter"
+		if err != nil {
+			msg = err.Error()
+		}
+
+		glog.Errorf("Route %s not admitted: %s", routeNameKey(route), msg)
+		p.recorder.RecordRouteRejection(route, "RouteNotAdmitted", msg)
+		return err
+	}
+
+	if err != nil {
+		msg := err.Error()
+		glog.Warningf("Route %s admitted with warnings: %s", routeNameKey(route), msg)
+		p.recorder.RecordRouteRejection(route, "RouteAdmissionWarning", msg)
+	}
+
+	if p.restrictOwnership && len(route.Spec.Host) > 0 {
+		switch eventType {
+		case watch.Added, watch.Modified:
+			if err := p.addRoute(route); err != nil {
+				glog.Errorf("Route %s not admitted: %s", routeNameKey(route), err.Error())
+				p.recorder.RecordRouteRejection(route, "SubdomainAlreadyClaimed", err.Error())
+				return err
+			}
+
+		case watch.Deleted:
+			p.claimedHosts.RemoveRoute(route.Spec.Host, route)
+			wildcardKey := routeapi.GetDomainForHost(route.Spec.Host)
+			p.claimedWildcards.RemoveRoute(wildcardKey, route)
+			p.blockedWildcards.RemoveRoute(wildcardKey, route)
+		}
+	}
+
+	return p.plugin.HandleRoute(eventType, route)
+}
+
+// HandleAllowedNamespaces limits the scope of valid routes to only those that match
+// the provided namespace list.
+func (p *HostAdmitter) HandleNamespaces(namespaces sets.String) error {
+	return p.plugin.HandleNamespaces(namespaces)
+}
+
+func (p *HostAdmitter) SetLastSyncProcessed(processed bool) error {
+	return p.plugin.SetLastSyncProcessed(processed)
+}
+
+// addRoute admits routes based on subdomain ownership - returns errors if the route is not admitted.
+func (p *HostAdmitter) addRoute(route *routeapi.Route) error {
+	// Find displaced routes (or error if an existing route displaces us)
+	displacedRoutes, err := p.displacedRoutes(route)
+	if err != nil {
+		err := fmt.Errorf("a route in another namespace holds host %s (%v)", route.Spec.Host, err)
```

Comment:
log the returned error (which has details about the other route), but don't report it to the user